### ISSUE TITLE
cant login in nspawn container via machinectl

### ISSUE THREAD

**Issue by shibumi** [systemd/systemd#2808]
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

> systemd version 229
### Used distribution

> Arch Linux
### Bugreport

I have started my container via `machinectl start work`.

```
work
           Since: Mon 2016-03-07 18:08:11 CET; 28min ago
          Leader: 8673 (systemd)
         Service: systemd-nspawn; class container
            Root: /var/lib/machines/work
           Iface: ve-work
            Unit: systemd-nspawn@work.service
                  ├─8669 /usr/bin/systemd-nspawn --quiet --keep-unit --boot --link-journal=try-guest --netw
                  ├─init.scope
                  │ └─8673 /usr/lib/systemd/systemd
                  └─system.slice
                    ├─systemd-journald.service
                    │ └─8741 /usr/lib/systemd/systemd-journald
                    └─console-getty.service
                      └─8778 /sbin/agetty --noclear --keep-baud console 115200 38400 9600 vt220

Mar 07 18:08:12 motoko systemd-nspawn[8669]: [1B blob data]
Mar 07 18:11:11 motoko systemd-nspawn[8669]: work login: systemd-logind.service: Failed at step STDIN spawn
Mar 07 18:11:12 motoko systemd-nspawn[8669]: systemd-logind.service: Failed at step STDIN spawning /usr/lib
Mar 07 18:11:12 motoko systemd-nspawn[8669]: systemd-logind.service: Failed at step STDIN spawning /usr/lib
Mar 07 18:11:12 motoko systemd-nspawn[8669]: systemd-logind.service: Failed at step STDIN spawning /usr/lib
Mar 07 18:11:12 motoko systemd-nspawn[8669]: systemd-logind.service: Failed at step STDIN spawning /usr/lib
Mar 07 18:23:47 motoko systemd-nspawn[8669]: systemd-tmpfiles-clean.service: Failed at step STDIN spawning 
Mar 07 18:29:13 motoko systemd[1]: Started Container work.
Mar 07 18:29:53 motoko systemd[1]: Started Container work.
Mar 07 18:30:39 motoko systemd[1]: Started Container work.
```

After this I wanted to login via `machinectl login work`. But I get the following Error:

```
Failed to get login PTY: Input/output error
```

**Comment by evverx** [comment]
```
systemd-logind.service: Failed at step STDIN spawning
```

Try to check `/dev/null`:

```
ls -l /dev/null
ls -l /var/lib/machines/work/dev/null
```

**Comment by shibumi** [comment]
@evverx the whole /dev directory is empty

**Comment by evverx** [comment]
hm, do you mean `ls -l /dev` on the host prints nothing?

**Comment by shibumi** [comment]
@evverx 
No I mean `/dev` in the container is empty.. 
Ok. After doing `systemd-nspawn -b -d work -n` it has booted up and i could login via `machinectl login work`.. seems like `machinectl start` is not initialising the container.

**Comment by evverx** [comment]
> seems like machinectl start is not initialising the container

Can you provide the output of `journalctl -b _PID=8669 -a`?

**Comment by shibumi** [comment]
@evverx 
https://paste.archlinux.de/SQ1t/
Btw.. I can't login via `machinectl login`. My last statement was wrong. I get the login dialog now, yes. But I cannot login because my password is always wrong.. seems like `machinectl login` is using `pts/0` but it should use `console` or `TTY1` or not?

**Comment by evverx** [comment]
> But I cannot login because my password is always wrong.

Can you login via `systemd-nspawn`?
I mean:

```
# systemd-nspawn -D your-cont -b
...
login: root
password:
```

**Comment by shibumi** [comment]
logging in via nspawn is working. machinectl login is not working.. I think it has something todo with `/etc/securetty`

**Comment by evverx** [comment]
> logging in via nspawn is working. machinectl login is not working

Ok. I understand.

BTW:
I was wrong about `ls -l /var/lib/machines/work/dev/null`
`nsenter --target 8673 --mount ls -l /dev/null` is the right command. Sorry.

**Comment by shibumi** [comment]
@evverx 
So is this a bug now or not? I mean normally should machinectl start initialize the container like nspawn.. but it's not working and this 'cant login'-issue is also strange

**Comment by evverx** [comment]
> should machinectl start initialize the container like nspawn.. 

`machinectl start` runs `systemd-nspawn@...service`. `systemd-nspawn@...service` runs `systemd-nspawn`.
So, `machinectl` should initialize the container like `nspawn`.

> So is this a bug now or not?

Maybe. I can't reproduce this.

>  this 'cant login'-issue is also strange

I agree.

**Comment by shibumi** [comment]
> So, machinectl should initialize the container like nspawn.

Yep and exactly this is not really working. I can create a second container if you want.. and i can try to reproduce this issue..

**Comment by evverx** [comment]
> I can create a second container if you want.. and i can try to reproduce this issue..

It would be great!

Let's check:

```
# recreate your container from scratch
machinectl start your-container
machinectl status your-container
...
          Leader: LEADER-PID (systemd)
...
           Unit: systemd-nspawn@work.service
                  ├─NSPAWN-PID /usr/bin/systemd-nspawn --quiet

...
journalctl -b _PID=NSPAWN-PID -a | grep 'Failed at step STDIN'
nsenter --target LEADER-PID --mount ls -l /dev
```

**Comment by evverx** [comment]
Updated: `journalctl -b _PID=NSPAWN-PID -a | grep 'Failed at step STDIN'` instead of `journalctl -b _PID=LEADER-PID`

**Comment by shibumi** [comment]
@evverx 
This time it's already failing at the beginning:

```
root@motoko /var/lib/machines # machinectl start testserver
root@motoko /var/lib/machines # machinectl status testserver
Could not get path to machine: No machine 'testserver' known
1 root@motoko /var/lib/machines # ls                                                                                                                                                                                                        :(
testserver  work
root@motoko /var/lib/machines # ls -la
total 16
drwx------  4 root root 4096 Mar  7 23:11 .
drwxr-xr-x 18 root root 4096 Mar  7 00:00 ..
drwxr-xr-x 16 root root 4096 Mar  7 23:08 testserver
drwxr-xr-x 16 root root 4096 Mar  7 18:28 work
root@motoko /var/lib/machines # machinectl list-images 
NAME       TYPE      RO  USAGE CREATED MODIFIED
testserver directory no  n/a   n/a     n/a     
work       directory no  n/a   n/a     n/a     

2 images listed.
```

`journalctl -b`

```
Mar 07 23:09:59 motoko systemd[1]: Starting Container testserver...
Mar 07 23:09:59 motoko systemd[425]: Started DWM statusbar.
Mar 07 23:09:59 motoko systemd-machined[7202]: New machine testserver.
Mar 07 23:09:59 motoko kernel: IPv6: ADDRCONF(NETDEV_UP): ve-testserver: link is not ready
Mar 07 23:09:59 motoko systemd-timesyncd[349]: Network configuration changed, trying to establish connection.
Mar 07 23:09:59 motoko systemd-nspawn[9672]: execv() failed: No such file or directory
Mar 07 23:09:59 motoko systemd[1]: Started Container testserver.
Mar 07 23:09:59 motoko polkitd[474]: Unregistered Authentication Agent for unix-process:9667:1927136 (system bus name :1.117, object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale en_US.UTF-8) (disconnected from bus)
Mar 07 23:10:00 motoko systemd[1]: systemd-nspawn@testserver.service: Main process exited, code=exited, status=1/FAILURE
Mar 07 23:10:00 motoko systemd[1]: systemd-nspawn@testserver.service: Unit entered failed state.
Mar 07 23:10:00 motoko systemd[1]: systemd-nspawn@testserver.service: Failed with result 'exit-code'.
Mar 07 23:10:00 motoko systemd[1]: Stopped Container testserver.
Mar 07 23:10:00 motoko systemd-machined[7202]: Machine testserver terminated.
Mar 07 23:10:04 motoko systemd-timesyncd[349]: Synchronized to time server 77.47.110.87:123 (2.arch.pool.ntp.org).
Mar 07 23:10:20 motoko systemd[425]: Started DWM statusbar.
Mar 07 23:11:20 motoko systemd[425]: Started DWM statusbar.
Mar 07 23:11:25 motoko polkitd[474]: Registered Authentication Agent for unix-process:9942:1935722 (system bus name :1.124 [/usr/bin/pkttyagent --notify-fd 5 --fallback], object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale 
Mar 07 23:11:25 motoko systemd[1]: Starting Container testserver...
Mar 07 23:11:25 motoko systemd[425]: Started DWM statusbar.
Mar 07 23:11:25 motoko systemd-machined[7202]: New machine testserver.
Mar 07 23:11:25 motoko systemd-nspawn[9947]: execv() failed: No such file or directory
Mar 07 23:11:25 motoko kernel: IPv6: ADDRCONF(NETDEV_UP): ve-testserver: link is not ready
Mar 07 23:11:25 motoko systemd[1]: Started Container testserver.
Mar 07 23:11:25 motoko systemd-timesyncd[349]: Network configuration changed, trying to establish connection.
Mar 07 23:11:25 motoko polkitd[474]: Unregistered Authentication Agent for unix-process:9942:1935722 (system bus name :1.124, object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale en_US.UTF-8) (disconnected from bus)
Mar 07 23:11:25 motoko systemd[1]: systemd-nspawn@testserver.service: Main process exited, code=exited, status=1/FAILURE
Mar 07 23:11:25 motoko systemd[1]: systemd-nspawn@testserver.service: Unit entered failed state.
Mar 07 23:11:25 motoko systemd[1]: systemd-nspawn@testserver.service: Failed with result 'exit-code'.
Mar 07 23:11:25 motoko systemd[1]: Stopped Container testserver.
Mar 07 23:11:25 motoko systemd-machined[7202]: Machine testserver terminated.
Mar 07 23:11:29 motoko systemd-timesyncd[349]: Synchronized to time server 77.47.110.87:123 (2.arch.pool.ntp.org).
```

```
root@motoko /var/lib/machines # systemd-nspawn -b -D testserver -n
Spawning container testserver on /var/lib/machines/testserver.
Press ^] three times within 1s to kill container.
execv() failed: No such file or directory
Container testserver failed with error code 1.
```

**Comment by evverx** [comment]
```
cd testserver
ls -l usr/lib/systemd/systemd lib/systemd/systemd sbin/init 
```

?

**Comment by shibumi** [comment]
```
2 root@motoko /var/lib/machines/testserver # ls                                                                                                                                                                                             :(
bin  boot  dev  etc  home  lib  lib64  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
root@motoko /var/lib/machines/testserver # ls -l usr/lib/systemd 
total 4
drwxr-xr-x 3 root root 4096 Mar  7 23:08 system
root@motoko /var/lib/machines/testserver # ls -l lib/systemd/systemd
ls: cannot access 'lib/systemd/systemd': No such file or directory
2 root@motoko /var/lib/machines/testserver # ls -l sbin/init                                                                                                                                                                                :(
ls: cannot access 'sbin/init': No such file or directory
```

**Comment by shibumi** [comment]
@evverx 
ahh sorry. I have just realized that I have created the container wrong.. I used mkarchroot instead of pacstrap. I will do it with pacstrap again

**Comment by shibumi** [comment]
@evverx https://paste.archlinux.de/gMbHy/ here is the right log

I can use `machinectl login testserver` now but I also can't login into my container because of `Login incorrect` although I have typed in the right root password.. login via nspawn is working

**Comment by evverx** [comment]
There is no `Failed at step STDIN`.
`systemd-logind` is running.
It's ok.

> I can use machinectl login testserver now

It's ok

> can't login into my container because of Login incorrect although I have typed in the right root password.. login via nspawn is working

Try `machinectl login` as non-root

**Comment by shibumi** [comment]
still not working. I need root for login in:

```
1 chris@motoko ~ % machinectl start testserver                                                                                                                                                                                              :(
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ===
Authentication is required to start 'systemd-nspawn@testserver.service'.
Authenticating as: root
Password: 
==== AUTHENTICATION COMPLETE ===
chris@motoko ~ % machinectl login testserver
==== AUTHENTICATING FOR org.freedesktop.machine1.login ===
Authentication is required to log into a local container.
Authenticating as: root
Password: 
==== AUTHENTICATION COMPLETE ===
Connected to machine testserver. Press ^] three times within 1s to exit session.

Arch Linux 4.4.3-1-ARCH (pts/0)

testserver login: root
Login incorrect

testserver login: root
Login incorrect

testserver login: 
Connection to machine testserver terminated.
```

**Comment by evverx** [comment]
Oh, sorry. I mean:

```
testserver login: some-non-root-user
...
```

**Comment by evverx** [comment]
Seems like https://github.com/systemd/systemd/issues/852

**Comment by shibumi** [comment]
As I thought.. adding `pts/0` to  `/etc/securetty` is fixing the login issue. Seems like this issue is: https://github.com/systemd/systemd/issues/852 but the first error is still creepy..
Why is `machinectl start` not initializing the container correctly? Why do I need to startup nspawn first and then with the second try with machinectl it works.

**Comment by poettering** [comment]
Friends don't let friends you pam_securetty.so. It's stupid and obsolete. Please ask your distro to remove it from their default PAM fragments.

But anyway, closing this, as this both a duplicate of #852 and also not a bug in systemd...

**Comment by AXGKl** [comment]
For what it's worth: selinux might also deny pty access at `machinectl login` and also `machinectl shell`.
Try if `setenforce=0` fixes it.
If so do, then do it the right way.

**Comment by Kytech** [comment]
I was encountering this issue on fedora server with SELinux set to enforced. Switching to permissive caused the issue to quit occurring. However, while in permissive mode, I discovered additional audit logs that were stopping this from working. Additionally, I was not able to narrow down all blocked actions until I temporarily disabled dontaudit rules. After doing this, I was able to use audit2allow to generate the needed policies.

To disable dontaudit rules:
`semodule -DB`
to reenable dontaudit rules when done:
`semodule -B`
